server.go

```diff
@@ -411,30 +411,23 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq, inbound
 
 	// Now that we've established a connection, create a peer, and
 	// it to the set of currently active peers.
-	peer, err := newPeer(conn, s, peerAddr, false)
+	p, err := newPeer(conn, connReq, s, peerAddr, inbound)
 	if err != nil {
 		srvrLog.Errorf("unable to create peer %v", err)
-		s.connMgr.Remove(peer.connReq.ID())
-		peer.Disconnect()
-		conn.Close()
+		p.Disconnect()
```

Comment:
The `Disconnect` function doesn't call `connMgr.Remove`, it calls `connMgr.Disconnect`. The former will tell it not to reconnect, while the latter continues to attempt to reconnect. If it keeps attempting to reconnect then we get node flapping as they'll get stuck in a loop. 

> We forgot to add check for nil, before removing connection

Yep, these issues can be fixed by retaining the diff I recommend above. Upon reflection, the `peer` struct itself really shouldn't be coupled with any information about the `connMgr` as we'd like to move it to a distinct package soon.